File: sdks/python/hatchet_sdk/worker/durable_eviction/manager.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from datetime import datetime, timedelta, timezone

from pydantic import BaseModel, ConfigDict

from hatchet_sdk.logger import logger
from hatchet_sdk.runnables.action import ActionKey
from hatchet_sdk.runnables.eviction import EvictionPolicy
from hatchet_sdk.worker.durable_eviction.cache import (
    DurableEvictionCache,
    DurableRunRecord,
)


class DurableEvictionConfig(BaseModel):
    model_config = ConfigDict(frozen=True)

    check_interval: timedelta = timedelta(seconds=1)
    """How often we try selecting an eviction candidate."""

    reserve_slots: int = 0
    """How many slots to reserve from capacity-based eviction decisions."""

    min_wait_for_capacity_eviction: timedelta = timedelta(seconds=10)
    """Avoid immediately evicting runs that have just entered a wait."""


DEFAULT_DURABLE_EVICTION_CONFIG = DurableEvictionConfig()


class DurableEvictionManager:
    def __init__(
        self,
        *,
        durable_slots: int,
        cancel_local: Callable[[ActionKey], None],
        request_eviction_with_ack: Callable[
            [ActionKey, DurableRunRecord], Awaitable[None]
        ],
        config: DurableEvictionConfig = DEFAULT_DURABLE_EVICTION_CONFIG,
        cache: DurableEvictionCache | None = None,
    ) -> None:
        self._durable_slots = durable_slots
        self._cancel_local = cancel_local
        self._request_eviction_with_ack = request_eviction_with_ack
        self._config = config
        self._cache = cache or DurableEvictionCache()

        self._task: asyncio.Task[None] | None = None
        self._lock = asyncio.Lock()

    @property
    def cache(self) -> DurableEvictionCache:
        return self._cache

    def start(self) -> None:
        # Lazy start (requires an event loop)
        if self._task and not self._task.done():
            return
        self._task = asyncio.create_task(self._run_loop())

    def stop(self) -> None:
        if self._task and not self._task.done():
            self._task.cancel()

    def register_run(
        self,
        key: ActionKey,
        *,
        step_run_id: str,
        eviction_policy: EvictionPolicy | None,
    ) -> None:
        self._cache.register_run(
            key,
            step_run_id,
            now=self._now(),
            eviction_policy=eviction_policy,
        )

    def unregister_run(self, key: ActionKey) -> None:
        self._cache.unregister_run(key)

    def mark_waiting(
        self,
        key: ActionKey,
        *,
        wait_kind: str,
        resource_id: str,
    ) -> None:
        self._cache.mark_waiting(
            key,
            now=self._now(),
            wait_kind=wait_kind,
            resource_id=resource_id,
        )

    def mark_active(self, key: ActionKey) -> None:
        self._cache.mark_active(key, now=self._now())

    async def _run_loop(self) -> None:
        interval = self._config.check_interval.total_seconds()

        try:
            while True:
                await asyncio.sleep(interval)
                await self._tick_safe()
        except asyncio.CancelledError:
            return

    async def _tick_safe(self) -> None:
        try:
            await self._tick()
        except Exception:
            logger.exception("DurableEvictionManager: error in eviction loop")

    async def _tick(self) -> None:
        # Only one eviction *cycle* at a time.
        #
        # Within a tick we drain all currently-eligible candidates
        async with self._lock:
            evicted_this_tick: set[ActionKey] = set()

            while True:
                key = self._cache.select_eviction_candidate(
                    now=self._now(),
                    durable_slots=self._durable_slots,
                    reserve_slots=self._config.reserve_slots,
                    min_wait_for_capacity_eviction=self._config.min_wait_for_capacity_eviction,
                )
                if key is None:
                    return

                # Safety: avoid infinite loops if cache repeatedly returns same key.
                if key in evicted_this_tick:
                    return
                evicted_this_tick.add(key)

                rec = self._cache.get(key)
                if rec is None:
                    continue

                if rec.eviction_policy is None:
                    continue

                logger.debug(
                    "DurableEvictionManager: evicting durable run "
                    f"task_run_external_id={rec.step_run_id} wait_kind={rec.wait_kind} "
                    f"resource_id={rec.wait_resource_id} ttl={rec.eviction_policy.ttl} "
                    f"capacity_allowed={rec.eviction_policy.allow_capacity_eviction}"
                )

                await self._request_eviction_with_ack(key, rec)

                self._cancel_local(key)
                self.unregister_run(key)

    async def evict_all_waiting(self) -> int:
        """Evict every currently-waiting durable run. Used during graceful shutdown."""
        self.stop()

        waiting = self._cache.get_all_waiting()
        evicted = 0

        for rec in waiting:
            if rec.eviction_policy is None:
                continue

            logger.debug(
                "DurableEvictionManager: shutdown-evicting durable run "
                f"task_run_external_id={rec.step_run_id} wait_kind={rec.wait_kind} "
                f"resource_id={rec.wait_resource_id}"
            )

            try:
                await self._request_eviction_with_ack(rec.key, rec)
            except Exception:
                logger.exception(
                    f"DurableEvictionManager: failed to send eviction for "
                    f"step_run_id={rec.step_run_id}"
                )
                continue

            self._cancel_local(rec.key)
            self.unregister_run(rec.key)
            evicted += 1

        return evicted

    def _now(self) -> datetime:
        return datetime.now(timezone.utc)

File: sdks/ruby/src/lib/hatchet/worker/durable_eviction/manager.rb
# frozen_string_literal: true

require "monitor"

require_relative "cache"

module Hatchet
  module WorkerRuntime
    module DurableEviction
      # Configuration for the background eviction loop.
      class DurableEvictionConfig
        # @return [Float] Seconds between eviction checks.
        attr_reader :check_interval
        # @return [Integer] Slots to reserve from capacity-eviction decisions.
        attr_reader :reserve_slots
        # @return [Float] Minimum seconds a run must have been waiting before it
        #   becomes eligible for capacity-based eviction.
        attr_reader :min_wait_for_capacity_eviction

        def initialize(check_interval: 1.0, reserve_slots: 0, min_wait_for_capacity_eviction: 10.0)
          @check_interval = check_interval
          @reserve_slots = reserve_slots
          @min_wait_for_capacity_eviction = min_wait_for_capacity_eviction
          freeze
        end
      end

      DEFAULT_DURABLE_EVICTION_CONFIG = DurableEvictionConfig.new

      # Orchestrates durable-task eviction.
      #
      # Runs a background thread that periodically selects an eviction candidate
      # from the cache, asks the server to evict it, and then interrupts the
      # local task thread.
      #
      # Mirrors :class:`hatchet_sdk.worker.durable_eviction.manager.DurableEvictionManager`.
      class DurableEvictionManager
        # @return [DurableEvictionCache]
        attr_reader :cache

        # @param durable_slots [Integer]
        # @param cancel_local [Proc] Called with the action key when the manager
        #   decides to evict a local run (invoked after the server ACK).
        # @param request_eviction_with_ack [Proc] Called with (action_key, DurableRunRecord)
        #   to send the eviction RPC to the server and block until acknowledged.
        # @param config [DurableEvictionConfig]
        # @param cache [DurableEvictionCache, nil]
        # @param logger [Logger, nil]
        def initialize(
          durable_slots:,
          cancel_local:,
          request_eviction_with_ack:,
          config: DEFAULT_DURABLE_EVICTION_CONFIG,
          cache: nil,
          logger: nil
        )
          @durable_slots = durable_slots
          @cancel_local = cancel_local
          @request_eviction_with_ack = request_eviction_with_ack
          @config = config
          @cache = cache || DurableEvictionCache.new
          @logger = logger

          @thread = nil
          @tick_monitor = Monitor.new
          @stopped = false
        end

        # Start the background eviction ticker. Idempotent.
        def start
          return if @thread&.alive?

          @stopped = false
          @thread = Thread.new { run_loop }
        end

        # Signal the background thread to stop. Does not join.
        def stop
          @stopped = true
          thread = @thread
          return unless thread&.alive?

          begin
            thread.wakeup
          rescue ThreadError
            nil
          end
        end

        # Register a new durable run invocation. Takes the current time from the
        # system clock.
        def register_run(key, step_run_id:, invocation_count:, eviction_policy:)
          @cache.register_run(
            key,
            step_run_id: step_run_id,
            invocation_count: invocation_count,
            now: now,
            eviction_policy: eviction_policy,
          )
        end

        # Unregister a durable run invocation.
        def unregister_run(key)
          @cache.unregister_run(key)
        end

        # Mark the run as waiting (increments the wait counter).
        def mark_waiting(key, wait_kind:, resource_id:)
          @cache.mark_waiting(key, now: now, wait_kind: wait_kind, resource_id: resource_id)
        end

        # Mark the run as active (decrements the wait counter).
        def mark_active(key)
          @cache.mark_active(key, now: now)
        end

        # Handle a server-initiated eviction notification for a stale invocation.
        def handle_server_eviction(step_run_id, invocation_count)
          key = @cache.find_key_by_step_run_id(step_run_id)
          return unless key

          rec = @cache.get(key)
          return if rec && rec.invocation_count != invocation_count

          @logger&.info(
            "DurableEvictionManager: server-initiated eviction for " \
            "step_run_id=#{step_run_id} invocation_count=#{invocation_count}",
          )
          evict_run(key)
        end

        # Evict every currently-waiting durable run. Used during graceful shutdown.
        #
        # @return [Integer] number of runs evicted
        def evict_all_waiting
          stop

          waiting = @cache.all_waiting
          evicted = 0

          waiting.each do |rec|
            rec.eviction_reason = DurableEvictionCache.build_eviction_reason(
              EvictionCause::WORKER_SHUTDOWN, rec,
            )

            @logger&.debug(
              "DurableEvictionManager: shutdown-evicting durable run " \
              "step_run_id=#{rec.step_run_id} wait_kind=#{rec.wait_kind} " \
              "resource_id=#{rec.wait_resource_id}",
            )

            begin
              @request_eviction_with_ack.call(rec.key, rec)
            rescue StandardError => e
              @logger&.error(
                "DurableEvictionManager: failed to send eviction for " \
                "step_run_id=#{rec.step_run_id}: #{e.class}: #{e.message}",
              )
            end

            # Always cancel locally even if the server ACK failed, so the
            # future settles and shutdown doesn't hang.
            evict_run(rec.key)
            evicted += 1
          end

          evicted
        end

        private

        def evict_run(key)
          @cancel_local.call(key)
          unregister_run(key)
        end

        def run_loop
          until @stopped
            sleep @config.check_interval
            break if @stopped

            tick_safe
          end
        rescue StandardError => e
          @logger&.error("DurableEvictionManager: run_loop exited: #{e.class}: #{e.message}")
        end

        def tick_safe
          tick
        rescue StandardError => e
          @logger&.error("DurableEvictionManager: error in eviction loop: #{e.class}: #{e.message}")
        end

        def tick
          @tick_monitor.synchronize do
            evicted_this_tick = []

            loop do
              key = @cache.select_eviction_candidate(
                now: now,
                durable_slots: @durable_slots,
                reserve_slots: @config.reserve_slots,
                min_wait_for_capacity_eviction: @config.min_wait_for_capacity_eviction,
              )
              return if key.nil?
              return if evicted_this_tick.include?(key)

              evicted_this_tick << key

              rec = @cache.get(key)
              next if rec.nil?
              next if rec.eviction_policy.nil?

              @logger&.debug(
                "DurableEvictionManager: evicting durable run " \
                "step_run_id=#{rec.step_run_id} wait_kind=#{rec.wait_kind} " \
                "resource_id=#{rec.wait_resource_id} ttl=#{rec.eviction_policy.ttl} " \
                "capacity_allowed=#{rec.eviction_policy.allow_capacity_eviction}",
              )

              @request_eviction_with_ack.call(key, rec)
              evict_run(key)
            end
          end
        end

        def now
          Time.now
        end
      end
    end
  end
end
